crates/node/src/roles/worker.rs

```diff
@@ -86,73 +85,140 @@ impl WorkerRole {
         self.bifrost.handle()
     }
 
-    pub async fn run(mut self, shutdown_watch: drain::Watch) -> Result<(), WorkerRoleError> {
+    pub fn worker_command_tx(&self) -> WorkerCommandSender {
+        self.worker.worker_command_tx()
+    }
+
+    pub fn storage_query_context(&self) -> &QueryContext {
+        self.worker.storage_query_context()
+    }
+
+    pub async fn run(
+        self,
+        _node_id: NodeId,
+        shutdown_watch: drain::Watch,
+    ) -> Result<(), WorkerRoleError> {
         let shutdown_signal = shutdown_watch.signaled();
 
         let (inner_shutdown_signal, inner_shutdown_watch) = drain::channel();
 
-        // Init the meta. This will reload the schemas in memory.
-        self.meta.init().await?;
+        // todo: only run subscriptions on node 0 once being distributed
+        let subscription_controller = Some(self.worker.subscription_controller_handle());
 
-        let worker_command_tx = self.worker.worker_command_tx();
-        let storage_query_context = self.worker.storage_query_context().clone();
-
-        let mut meta_handle = tokio::spawn(
-            self.meta
-                .run(inner_shutdown_watch.clone(), worker_command_tx.clone()),
-        );
-        let mut admin_handle = tokio::spawn(self.admin.run(
-            inner_shutdown_watch.clone(),
-            worker_command_tx,
-            Some(storage_query_context),
-        ));
+        let mut component_set = JoinSet::new();
 
         // Ensures bifrost has initial metadata synced up before starting the worker.
-        let mut bifrost_handle = self.bifrost.start(inner_shutdown_watch.clone()).await?;
+        let mut bifrost_join_handle = self.bifrost.start(inner_shutdown_watch.clone()).await?;
 
-        let mut worker_handle = tokio::spawn(self.worker.run(inner_shutdown_watch));
+        component_set.spawn(
+            self.worker
+                .run(inner_shutdown_watch)
+                .map_ok(|_| "worker")
+                .map_err(WorkerRoleError::Worker),
+        );
+        component_set.spawn(
+            Self::reload_schemas(
```

Comment:
I assumed we will need to perform a blocking initial schema fetch _before_ starting the worker. Do we have a risk of partition processors failing/ignoring invocations because they can't see the schema on startup, or if it takes time? I'm worried about the risk of data loss if PP drops an invocation because it thinks the service have been removed (vs. that we are running with old Meta information).

Reply:
The PP won't drop any invocations it does not know about. What can happen is that the ingress rejects for some time new invocations until it learns about them. What can also happen, and this is probably not so nice, is that the PP will try to invoke some invocations for which it does not know the deployment information. This would then lead to a couple of retries which causes some noise in the logs. To prevent this from happening trying to fetch the schema before starting the worker is probably a good idea. This won't solve the problem in all cases but in most.

Reply:
This sounds like a fragile design assumption that might not stand the test of time. The good news is that it won't cause problems at the moment, but I think we can have a more robust approach with the following design:
- Schema is versioned.
- When schema is updated on an admin node, the admin node can inform all nodes with "ingress" role with the update. This is either a notification of the higher version number, or a full push of the new schema object. Ideally, we finish up the ingress version notification _before_ we respond to user-requested schema changes, this is to achieve monotonic read.
- Ingress writes its last-seen schema version in all messages going through bifrost or in RPC to workers. A worker processing an invocation message will compare the schema version, if it's running with an older version, it pulls the new schema before moving forward. If the worker has the same schema as the incoming message (whether from another PP or ingress) it can _more_ confidently reject invocations if we need to do that in the future.

Reply:
Sending the schema information version with the invocation messages is a good idea to achieve monotonic reads :-)